# Practice Session 04: Basket analysis

Author: <font color="blue">Manvir Kaur Singh</font>

E-mail: <font color="blue">manvir.kaur01@estudiant.upf.edu</font>

Date: <font color="blue">21/10/2023</font>

In [1]:
pip install apyori

Note: you may need to restart the kernel to use updated packages.


In [2]:
import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd  
import csv
import gzip
from apyori import apriori

# 1. Playing with apyori

In [3]:
# LEAVE AS-IS

def print_apyori_output (association_results, info=False, info_key=False):
    for relation_record in association_results:
        itemset = list(relation_record.items)
        
        # Consider only itemsets of two elements
        if len(itemset) > 1: 
        
            print("Rules involving itemset %s" % itemset)
            support = relation_record.support

            for rules in relation_record.ordered_statistics:
                antecedent = list(rules.items_base)
                consequent = list(rules.items_add)
                
                if info_key:
                    antecedent = [info.loc[x][info_key] for x in antecedent]
                    consequent = [info.loc[x][info_key] for x in consequent]
                
                confidence = rules.confidence
                lift = rules.lift

                print("%s => %s (support=%.4f, confidence=%.2f, lift=%.2f)" %
                      (antecedent, consequent, support, confidence, lift))
            print()

In [4]:
transactions = [
    ['beer', 'chips', 'nuts', 'olives'],
    ['beer', 'chips', 'olives'],
    ['chips', 'nuts' ],
    ['chips', 'olives'],
    ['beer', 'nuts' ],
    ['chips'],
    ['nuts', 'olives'],
    ['beer', 'nuts'],
    ['beer', 'chips', 'olives'], 
    ['beer', 'nuts', 'olives'], 

]
results = list(apriori(transactions, min_support=0.3, min_confidence=0.75, min_lift=1.0))

print_apyori_output(results)

Rules involving itemset ['beer', 'olives', 'chips']
['beer', 'chips'] => ['olives'] (support=0.3000, confidence=1.00, lift=1.67)



<font size="+1" color="red">Replace this cell with your own example of transactions (at least 20 transactions) and execution of the apriori algorithm, in which you should obtain at least ONE and at most THREE rules.</font>

In [5]:
transactions = [
    ['orangejuice', 'apple', 'toast'],
    ['nuts', 'milk', 'coffee'],
    ['orangejuice', 'toast' , 'coffee'],
    ['apple', 'toast', 'coffee'],
    ['milk', 'nuts', 'coffee' ],
    ['orangejuice', 'apple', 'toast', 'coffee'],
    ['nuts', 'milk'],
    ['orangejuice', 'coffee'],
    ['toast', 'coffee'],
    ['milk'],
    ['milk', 'apple'],
    ['nuts', 'toast', 'coffee'],
    ['toast' , 'coffee'],
    ['apple', 'toast'],
    ['nuts', 'coffee' ],
    ['apple', 'toast', 'coffee'],
    ['nuts'],
    ['coffee'],
    ['toast', 'coffee','orangejuice'],
    ['toast', 'nuts'],
    ['orangejuice', 'apple', 'coffee']
    

]
results = list(apriori(transactions, min_support=0.3, min_confidence=0.6, min_lift=1.0))

<font size="+1" color="red">Replace this cell with (1) a printout of the rules you have obtained, and (2) for each of those rules, indicate clearly how the support, confidence, and lift is calculated. Do not merely repeat the formula: indicate how each number is computed based on the transactions you provided, as if you were trying to verify that the numbers are correct.</font>

In [6]:
print_apyori_output(results)

Rules involving itemset ['coffee', 'toast']
['toast'] => ['coffee'] (support=0.3810, confidence=0.73, lift=1.09)



Calculations for the Rule: ['toast'] => ['coffee']
SUPPORT:

SUPPORT('toast' => 'coffee') = (NUM OF TRANSACTIONS CONTAINING 'TOAST' AND 'COFFEE')/(TOTAL NUM OF TRANSACTIONS) = 8/21 = 0.381

CONFIDENCE:

CONFIDENCE('toast' => 'coffee') = (SUPPORT('toast' => 'coffee'))/(SUPPORT('toast')) = (8/21)/(11/21)= 8/11 = 0.73

LIFT:

LIFT('toast' => 'coffee') = (CONFIDENCE('toast' => 'coffee'))/(SUPPORT('coffee'))= (8/11)/(14/21) = 1.09



Here, the lift value indicates that the likelihood of buying 'toast' and 'coffee' together is 1.09 times higher than if they were statistically independent.

# 2. Load and prepare the shopping baskets

In [7]:
# LEAVE AS-IS

# File names
INPUT_PRODUCTS = "instacart-products.csv"
INPUT_TRANSACTIONS = "instacart-transactions.csv.gz"

# Read into a dataframe
products = pd.read_csv(INPUT_PRODUCTS, delimiter=",")

# Set product_id as index, and drop column aisle_id
products = products.set_index('product_id').drop(columns=['aisle_id'])

products.head(100)

product_name  department_id
product_id                                                                  
1                                  Chocolate Sandwich Cookies             19
2                                            All-Seasons Salt             13
3                        Robust Golden Unsweetened Oolong Tea              7
4           Smart Ones Classic Favorites Mini Rigatoni Wit...              1
5                                   Green Chile Anytime Sauce             13
...                                                       ...            ...
96                    Sprinklez Confetti Fun Organic Toppings             13
97                                Organic Chamomile Lemon Tea              7
98                                  2% Yellow American Cheese             16
99                                Local Living Butter Lettuce              4
100                   Peanut Butter & Strawberry Jam Sandwich              1

[100 rows x 2 columns]

## 2.1. Select by department

In [8]:
# LEAVE AS-IS

DEPT_BAKERY = 3
DEPT_VEGGIES = 4
DEPT_ALCOHOL = 5
DEPT_WORLD = 6
DEPT_DRINKS = 7
DEPT_PETS = 8
DEPT_PHARMACY = 11
DEPT_CLEANING = 17
DEPT_BABIES = 18

<font size="+1" color="red">Replace this cell with your code for *select_from_departments*.</font>

In [9]:
def select_from_departments(products, product_ids, department_ids):
    s_products = []
    for p_id in product_ids:
        if p_id in products.index and products.loc[p_id].department_id in department_ids:
            s_products.append(p_id)
    return s_products

<font size="+1" color="red">Replace this cell with code to test your function with three different test cases. Each test case is a list of items and a list of 1, 2, or 3 departments.</font>

In [10]:
products_ids = [22, 26, 45, 54, 57, 71, 111, 112]
department_ids = [DEPT_BAKERY, DEPT_CLEANING]

products_data = {
    'product_name': ['Fresh Breath Oral Rinse Mild Mint', 'Fancy Feast Trout Feast Flaked Wet Cat Food', 'European Cucumber', '24/7 Performance Cat Litter', 'Flat Toothpicks', 'Ultra 7 Inch Polypropylene Traditional Plates', 'Fabric Softener, Geranium Scent', 'Hot Tomatillo Salsa'],
    'department_id': [11, 8, 4, 8, 17, 17, 17, 13]
}

products_1 = pd.DataFrame(products_data, index=products_ids)

print(select_from_departments(products_1, products_ids, department_ids))



products_ids_2 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
departments_ids_2 = [DEPT_VEGGIES, DEPT_DRINKS]

products_data_2 = {
    'product_name': ['Apple', 'Orange Juice', 'Carrot', 'Soda', 'Lettuce', 'Milk', 'Tomato', 'Water', 'Banana', 'Cucumber'],
    'department_id': [DEPT_VEGGIES, DEPT_DRINKS, DEPT_VEGGIES, DEPT_DRINKS, DEPT_VEGGIES, DEPT_DRINKS, DEPT_VEGGIES, DEPT_DRINKS, DEPT_VEGGIES, DEPT_VEGGIES]
}

products_2 = pd.DataFrame(products_data_2, index=products_ids_2)

print(select_from_departments(products_2, products_ids_2, departments_ids_2))




products_ids_3 = [111, 101, 102, 103, 104, 105, 106, 107, 108]
departments_ids_3 = [DEPT_ALCOHOL, DEPT_BABIES]

products_data_3 = {
    'product_name': ['Baby Bottle', 'Diapers', 'Baby Food', 'Baby Shampoo', 'Pacifier', 'Baby Wipes', 'Teething Toy', 'Baby Lotion', 'Baby Clothes'],
    'department_id': [DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES, DEPT_BABIES]
}

products_3 = pd.DataFrame(products_data_3, index=products_ids_3)

print(select_from_departments(products_3, products_ids_3, departments_ids_3))


[57, 71, 111]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[111, 101, 102, 103, 104, 105, 106, 107, 108]


## 2.2. Read and filter transactions

<font size="+1" color="red">Replace this cell with your code to read transactions, keeping only items in DEPT_CLEANING. Remember to stop after keeping 5000 transactions.</font>

In [11]:
# Open a compressed file

transactions = []
trans_read = 0
trans_stored = 0
with gzip.open(INPUT_TRANSACTIONS, "rt") as inputfile:
    
    # Create a CSV reader
    reader = csv.reader(inputfile, delimiter=",")
    
    # Iterate through the CSV file
    for row in reader:
        
        # Convert to integers
        items = [int(x) for x in row]
        
        if any(products.loc[item].department_id == DEPT_CLEANING for item in items):
            transactions.append(items)
            trans_stored += 1
            
        trans_read += 1
        
        if trans_read % 1000 ==0:
            print (f"Transaction read: {trans_read}, Transaction stored: {trans_stored}")
        
        if trans_stored >=5000:
            break

Transaction read: 1000, Transaction stored: 158
Transaction read: 2000, Transaction stored: 311
Transaction read: 3000, Transaction stored: 460
Transaction read: 4000, Transaction stored: 598
Transaction read: 5000, Transaction stored: 745
Transaction read: 6000, Transaction stored: 902
Transaction read: 7000, Transaction stored: 1067
Transaction read: 8000, Transaction stored: 1206
Transaction read: 9000, Transaction stored: 1373
Transaction read: 10000, Transaction stored: 1515
Transaction read: 11000, Transaction stored: 1670
Transaction read: 12000, Transaction stored: 1807
Transaction read: 13000, Transaction stored: 1951
Transaction read: 14000, Transaction stored: 2102
Transaction read: 15000, Transaction stored: 2245
Transaction read: 16000, Transaction stored: 2384
Transaction read: 17000, Transaction stored: 2543
Transaction read: 18000, Transaction stored: 2692
Transaction read: 19000, Transaction stored: 2840
Transaction read: 20000, Transaction stored: 2974
Transaction rea

## 2.3. Extract association rules and comment on them (DEPT_CLEANING)

<font size="+1" color="red">Replace this cell with your code to extract association rules from the read transactions (DEPT_CLEANING).</font>

In [12]:
results = list(apriori(transactions, min_support=0.01, min_confidence=0.35, min_lift=0.1))
print_apyori_output(results, products, 'product_name')


Rules involving itemset [13176, 5876]
['Organic Lemon'] => ['Bag of Organic Bananas'] (support=0.0104, confidence=0.37, lift=3.25)

Rules involving itemset [47209, 5876]
['Organic Lemon'] => ['Organic Hass Avocado'] (support=0.0100, confidence=0.36, lift=5.80)

Rules involving itemset [13176, 27966]
['Organic Raspberries'] => ['Bag of Organic Bananas'] (support=0.0158, confidence=0.39, lift=3.40)

Rules involving itemset [13176, 47209]
['Organic Hass Avocado'] => ['Bag of Organic Bananas'] (support=0.0222, confidence=0.36, lift=3.15)



<font size="+1" color="red">Replace this cell with a brief commentary on what you would recommend to the shopping app considering the extracted association rules.</font>

Based on the extracted association rules, we can see that certain organic products are strongly associated with each other. Specifically, products like 'Organic Lemon', 'Organic Hass Avocado', and 'Organic Raspberries' show strong relationships with 'Bag of Organic Bananas'. So the shopping app can give disscount of bag of organic bananas whenever they buy organic raspberries, organic lemon or organic hass avocado.

## 2.4. Extract association rules and comment on them (other departments)

<font size="+1" color="red">Replace this cell with code to select a different set of departments (at least two, not DEPT_CLEANING) and extract transactions again. Avoid replicating code when possible.</font>

In [13]:
transactions = []
trans_read = 0
trans_stored = 0
with gzip.open(INPUT_TRANSACTIONS, "rt") as inputfile:
    
    # Create a CSV reader
    reader = csv.reader(inputfile, delimiter=",")
    
    # Iterate through the CSV file
    for row in reader:
        
        # Convert to integers
        items = [int(x) for x in row]
        
        if any(products.loc[item].department_id == DEPT_VEGGIES for item in items):
            transactions.append(items)
            trans_stored += 1
            
        trans_read += 1
        
        if trans_read % 1000 ==0:
            print (f"Transaction read: {trans_read}, Transaction stored: {trans_stored}")
        
        if trans_stored >=5000:
            break

Transaction read: 1000, Transaction stored: 773
Transaction read: 2000, Transaction stored: 1518
Transaction read: 3000, Transaction stored: 2267
Transaction read: 4000, Transaction stored: 3018
Transaction read: 5000, Transaction stored: 3760
Transaction read: 6000, Transaction stored: 4493


In [14]:
results = list(apriori(transactions, min_support=0.01, min_confidence=0.35, min_lift=0.1))
print_apyori_output(results, products, 'product_name')

Rules involving itemset [24852, 28204]
['Organic Fuji Apple'] => ['Banana'] (support=0.0144, confidence=0.38, lift=1.87)

Rules involving itemset [30489, 24852]
['Original Hummus'] => ['Banana'] (support=0.0106, confidence=0.36, lift=1.77)

Rules involving itemset [45066, 24852]
['Honeycrisp Apple'] => ['Banana'] (support=0.0158, confidence=0.41, lift=2.06)

Rules involving itemset [49683, 24852]
['Cucumber Kirby'] => ['Banana'] (support=0.0132, confidence=0.35, lift=1.76)



<font size="+1" color="red">Replace this cell with your commentary on the obtained rules.</font>

The generated association rules reveal interesting patterns in customer purchasing behavior, particularly regarding the interaction between fresh fruits and specific items like 'Banana'. 

<font size="+2" color="#003300">I hereby declare that, except for the code provided by the course instructors, all of my code, report, and figures were produced by myself.</font>